### ISSUE TITLE
Memory Leak with Docker GPU

### ISSUE THREAD

**Issue by vwxyzjn** [pytorch/pytorch#38910]
## 🐛 Bug

I built a custom image based on `pytorch/pytorch:1.5-cuda10.1-cudnn7-runtime`, but it has memory leak once ran using docker with gpus. As an example, this ran (https://app.wandb.ai/cleanrl/cleanRL/runs/1giu2q8i?workspace=user-costa-huang) has its memory risen from 9% to 30% of 64 GB memory in minutes. In comparison, I rolled back to build the custom image based on `pytorch/pytorch:1.4-cuda10.1-cudnn7-runtime` and there is no such issues.

![image](https://user-images.githubusercontent.com/5555347/82630987-680f1100-9bc2-11ea-8ef2-cedbe0452310.png)

## To Reproduce

Obtain a wandb key from https://www.wandb.com/ and run the following

```
docker run --gpus all -d --cpuset-cpus="0" -e WANDB={YOUR_WANDB_KEY} vwxyzjn/cleanrl:1.5-cuda10.1-cudnn7-runtime-cleanrl-0.2.1 python dqn_atari_visual.py --gym-id BreakoutNoFrameskip-v4 --total-timesteps 10000000 --wandb-project-name cleanrl.benchmark --wandb-entity cleanrl --prod-mode --capture-video --seed 2
```
Should reproduce the results where the memory blows up about 10 mins in. While the following script that uses the previous docker image of pytorch 1.4 should not have its memory blow up
```
docker run --gpus all -d --cpuset-cpus="0" -e WANDB={YOUR_WANDB_KEY} vwxyzjn/cleanrl:1.4-cuda10.1-cudnn7-runtime-cleanrl-0.2.1 python dqn_atari_visual.py --gym-id BreakoutNoFrameskip-v4 --total-timesteps 10000000 --wandb-project-name cleanrl.benchmark --wandb-entity cleanrl --prod-mode --capture-video --seed 2
```

And this is the Dockerfile

```
FROM pytorch/pytorch:1.5-cuda10.1-cudnn7-runtime

RUN apt-get update && \
    apt-get -y install xvfb ffmpeg git build-essential
RUN pip install gym[box2d,atari] pybullet==2.7.9
RUN git clone https://github.com/vwxyzjn/cleanrl && \
    cd cleanrl && pip install -e .
RUN apt-get -y install python-opengl
RUN pip install opencv-python
RUN pip install seaborn pandas

WORKDIR /workspace/cleanrl/cleanrl

COPY entrypoint.sh /usr/local/bin/
RUN chmod 777 /usr/local/bin/entrypoint.sh
ENTRYPOINT ["/usr/local/bin/entrypoint.sh"]
CMD python ppo2_continuous_action.py --capture-video --total-timesteps 200
```

## Expected behavior

The following image is the expected behavior, which is produced by the previous pytorch 1.4 docker image (https://app.wandb.ai/cleanrl/cleanrl.benchmark/runs/3az3ndy1?workspace=user-costa-huang)
![image](https://user-images.githubusercontent.com/5555347/82631048-8ecd4780-9bc2-11ea-84bc-fc51dbded0fb.png)

In particular, the memory didn't blow up.

## Environment

The experiments are ran using AWS batch at a GPU instance, and the docker image is based on `pytorch/pytorch:1.5-cuda10.1-cudnn7-runtime` that produces the memory explosion.

**Comment by ezyang** [comment]
If you could give us some high level characteristics about the model in question that would be helpful in pinning things down. I notice you are setting cuDNN to be deterministic so I speculatively added the cudnn label here. Another helpful thing might be to check if master also leaks in the same way.

**Comment by vwxyzjn** [comment]
On a high-level, the model is a Deep Q learning algorithm that has the following core structure. 

```python
for global_step in range(args.total_timesteps):
    # ALGO LOGIC: put action logic here
    epsilon = linear_schedule(args.start_e, args.end_e, args.exploration_fraction*args.total_timesteps, global_step)
    logits = q_network.forward(obs.reshape((1,)+obs.shape))
    if args.capture_video:
        env.set_q_values(logits.tolist())
    if random.random() < epsilon:
        action = env.action_space.sample()
    else:
        action = torch.argmax(logits, dim=1).tolist()[0]

    # TRY NOT TO MODIFY: execute the game and log data.
    next_obs, reward, done, _ = env.step(action)
    episode_reward += reward

    # ALGO LOGIC: training.
    rb.put((obs, action, reward, next_obs, done))
    if global_step > args.learning_starts and global_step % args.train_frequency == 0:
        s_obs, s_actions, s_rewards, s_next_obses, s_dones = rb.sample(args.batch_size)
        with torch.no_grad():
            target_max = torch.max(target_network.forward(s_next_obses), dim=1)[0]
            td_target = torch.Tensor(s_rewards).to(device) + args.gamma * target_max * (1 - torch.Tensor(s_dones).to(device))
        old_val = q_network.forward(s_obs).gather(1, torch.LongTensor(s_actions).view(-1,1).to(device)).squeeze()
        loss = loss_fn(td_target, old_val)
        writer.add_scalar("losses/td_loss", loss, global_step)

        # optimize the midel
        optimizer.zero_grad()
        loss.backward()
        nn.utils.clip_grad_norm_(list(q_network.parameters()), args.max_grad_norm)
        optimizer.step()

        # update the target network
        if global_step % args.target_network_frequency == 0:
            target_network.load_state_dict(q_network.state_dict())

    # TRY NOT TO MODIFY: CRUCIAL step easy to overlook 
    obs = next_obs

    if done:
        # TRY NOT TO MODIFY: record rewards for plotting purposes
        if env.was_real_done:
            print(f"global_step={global_step}, episode_reward={real_episode_reward}")
            writer.add_scalar("charts/episode_reward", real_episode_reward, global_step)
            writer.add_scalar("charts/epsilon", epsilon, global_step)
            real_episode_reward = 0
        else:
            real_episode_reward += episode_reward
        obs, episode_reward = env.reset(), 0

env.close()
writer.close()
```

The code is self-contained in a single file at https://app.wandb.ai/cleanrl/cleanRL/runs/1giu2q8i/code?workspace=user-costa-huang, and I expect you to just be able to run after downloading it. 

Unfortunately I am not very familiar with how to build the images from pytorch master, so it might be difficult for me to check if master also leaks.

Thanks.

**Comment by seemethere** [comment]
Documentation for building from master can be found here: https://github.com/pytorch/pytorch#building-the-image-yourself

**Comment by xwang233** [comment]
I believe the memory keeps increasing is because this line in your script
```
rb.put((obs, action, reward, next_obs, done))
```
which keeps pushing data into a `collections.deque()`. The default `max_len` (`buffer_size` in your script) of the deque is 1000000 in your script, so it will keep increasing until the 1 million-th step. 

By commenting out that line, everything else seems well and the memory keeps the same. This issue is not docker related nor pytorch related. So I suggest close it.

**Comment by vwxyzjn** [comment]
@xwang233 hi thanks for following up. I don’t believe that is the issue.

Running the following command blows up the memory with pytorch1.5 docker image
`docker run --gpus all -d --cpuset-cpus="0" -e WANDB={YOUR_WANDB_KEY} vwxyzjn/cleanrl:1.5-cuda10.1-cudnn7-runtime-cleanrl-0.2.1 python dqn_atari_visual.py --gym-id BreakoutNoFrameskip-v4 --total-timesteps 10000000 --wandb-project-name cleanrl.benchmark  --prod-mode --capture-video --seed 2`

While running the same exact file using pytorch 1.4 docker image does not blow up the memory `docker run --gpus all -d --cpuset-cpus="0" -e WANDB={YOUR_WANDB_KEY} vwxyzjn/cleanrl:1.4-cuda10.1-cudnn7-runtime-cleanrl-0.2.1 python dqn_atari_visual.py --gym-id BreakoutNoFrameskip-v4 --total-timesteps 10000000 --wandb-project-name cleanrl.benchmark --prod-mode --capture-video --seed 2`

**Comment by xwang233** [comment]
@vwxyzjn Oh, I see what you mean. Can you do two plots of system memory vs training steps instead of training times?

**Comment by vwxyzjn** [comment]
@xwang233 unfortunately the weights and biases Dashboard does not allow me to create such plots directly. However, if you read the first figure, you will find the training steps (global_steps) grows linearly with training time, while the memory usage grows exponentially with training time. 

In comparison, the memory usage grows linearly with training time in pytorch1.4 docker images as shown here https://app.wandb.ai/cleanrl/cleanrl/runs/3az3ndy1?workspace=user-costa-huang

**Comment by mattip** [comment]
Testing with latest HEAD.

If I instrument the file to report the rss and vss memory with psutils and running the code with `--buffer-size=1000`, I can confirm that the memory growth stops when the dequeue buffer is full (in each iteration `new_mem - old_mem` remains `0`. See the details for the diff to the original code. 

However that was when running with only `--buffer-size=1000`. The docker command runs with 
```
--gym-id BreakoutNoFrameskip-v4 
--total-timesteps 10000000 
--wandb-project-name cleanrl.benchmark 
--prod-mode 
--capture-video 
--seed 2
```

When I add the `--capture-video` argument only, memory still is constant after the dequeue buffer is full. Maybe the problem is some interaction with [`wandb`](https://pypi.org/project/wandb/)?
A quick scan of the wandb client issue tracker shows wandb/client#354 which was closed in version 0.8.2, released in June 2019. @vwxyzjn what version of wandb were you using in each case? 


<details>

```
--- /tmp/code_original.py	2021-01-18 19:46:58.079515419 +0200
+++ /tmp/code_modified.py	2021-01-18 19:52:13.595189108 +0200
@@ -1,3 +1,4 @@
+
 # https://github.com/facebookresearch/torchbeast/blob/master/torchbeast/core/environment.py
 
 import numpy as np
@@ -6,6 +7,8 @@
 from gym import spaces
 import cv2
 cv2.ocl.setUseOpenCL(False)
+import psutil
+ps = psutil.Process()
 
 
 class NoopResetEnv(gym.Wrapper):
@@ -520,6 +523,7 @@
 # important to note that because `EpisodicLifeEnv` wrapper is applied,
 # the real episode reward is actually the sum of episode reward of 5 lives
 real_episode_reward = 0
+old_mem = [0, 0]
 for global_step in range(args.total_timesteps):
     # ALGO LOGIC: put action logic here
     epsilon = linear_schedule(args.start_e, args.end_e, args.exploration_fraction*args.total_timesteps, global_step)
@@ -563,12 +567,18 @@
         # TRY NOT TO MODIFY: record rewards for plotting purposes
         if env.was_real_done:
             print(f"global_step={global_step}, episode_reward={real_episode_reward}")
             writer.add_scalar("charts/episode_reward", real_episode_reward, global_step)
             writer.add_scalar("charts/epsilon", epsilon, global_step)
             real_episode_reward = 0
+            pmem = ps.memory_info()
+            new_mem = [pmem.rss, pmem.vms]
+            delta = [n-o for n,o in zip(new_mem, old_mem)]
+            print(global_step, f'delta:{delta}')
+            old_mem = new_mem
         else:
             real_episode_reward += episode_reward
         obs, episode_reward = env.reset(), 0

```

**Comment by vwxyzjn** [comment]
Hi @mattip, this is probably not a wandb issue. As logged by the `requirements.txt` by both run, (https://wandb.ai/cleanrl/cleanRL/runs/1giu2q8i/files/requirements.txt, https://wandb.ai/cleanrl/cleanrl/runs/3az3ndy1/files/requirements.txt), the version is `wandb==0.8.36`, yet one of the run had that memory leak.

**Comment by mattip** [comment]
But without wandb the memory use is constant (once the dequeue buffer is full), so if not an issue directly with wandb, it is some interaction between wandb and the pytorch version.

**Comment by vwxyzjn** [comment]
Thanks for following up btw. Quick question. Have you tried to run the scripts using `buffer_size=1000000`? I recall the memory didn't start to blow up until 15 mins into the run.

> But without wandb the memory use is constant (once the dequeue buffer is full), so if not an issue directly with wandb, it is some interaction between wandb and the pytorch version.
That is possible. I'd be happy to test it out on the latest pytorch image in a couple days. Maybe it is no longer an issue.

**Comment by mattip** [comment]
I tried without supplying a `buffer-size` (and without wandb).  Memory use rose linearly until it got to about 90% of my 16GB of RAM, and then it began behaving strangely. Perhaps the dequeue buffer was unable to realloc.

I see a worrying pause in the `global_step` at around 12 minutes in the top right graph. That is strange, since the global step is a loop counter. Why should there be a 1 minute pause? It may be helpful to try to disentangle the interactions. When you get a chance could you try
- without wandb (but with the instrumentation I added to the script in my first comment
- with wandb and with a limited `buffer-size`?

**Comment by malfet** [comment]
Removing milestone tag as there does not seem to be any progress since Jan 2021
